### DISCUSSION TITLE
finite element method works in moose

### DISCUSSION THREAD

**Discussion by Joseph-0123** [idaholab/moose#22333] [Q&A General] [answered]
Dear all, could you please help with the following questions：

1. how does the finite element method works in moose? for example, if we have a 1D domain with 5 global nodes, now we need to solve two different equations such as the mass equation and the Navier stokes equation. In the end, we will sum up all the residuals from two equations together such that residual from eq 1+residual from eq 2=0 ?

2. Similarly, if we have momentum equations in 3 directions, will we sum up all residuals from 3 directions and let it =0?

3. In the Navier stokes equation, we may have a double derivative of velocity in the viscous term, after we multiply the test function, we still have a gradient of velocity which is one variable I want to solve. whenever I use _grad_u to represent it, moose tells me some error. Is it because the grad_u (gradient of velocity) is a tensor since velocity is a tensor?

Thanks a lot.

J

**Comment by GiudGiud** [comment, marked as answer]
Hello

No sum of residuals involved. The residuals are a field for each equation
Maybe this documentation page will help
https://mooseframework.inl.gov/source/systems/NonlinearSystem.html#newtons_method

_grad_u is of the type you specified in the header of the source file. 
If u is the whole velocity vector, then yes it s a tensor.
If u is just a component, then grad_u can be a vector

Guillaume

**Reply by Joseph-0123** [reply]
> Hello
> 
> No sum of residuals involved. The residuals are a field for each equation Maybe this documentation page will help https://mooseframework.inl.gov/source/systems/NonlinearSystem.html#newtons_method
> 
> _grad_u is of the type you specified in the header of the source file. If u is the whole velocity vector, then yes it s a tensor. If u is just a component, then grad_u can be a vector
> 
> Gyukkayne

Thanks for your answers, Gyukkayne. I learned an [INSADMaterial](https://github.com/idaholab/moose/blob/next/modules/navier_stokes/src/materials/INSADMaterial.C) in [navier_stokes](https://github.com/idaholab/moose/tree/next/modules/navier_stokes) module of MOOSE.
```
_INSADMaterial::INSADMaterial(const InputParameters & parameters)
  : Material(parameters),
    _velocity(adCoupledVectorValue("velocity")),
    _grad_velocity(adCoupledVectorGradient("velocity")),_
```
Here. the _velocity and _grad_velocity are all Vectors. It may not match the 'If u is the whole velocity vector, then yes it s a tensor.' Is it correct ? Thanks. 

J

**Reply by GiudGiud** [reply]
See the doxygen for that adCoupledVectorGradient
```
const ADVectorVariableGradient & | adCoupledVectorGradient (const std::string &var_name, unsigned int comp=0) const
```

Then clicking on the return type:
https://mooseframework.inl.gov/docs/doxygen/moose/MooseTypes_8h.html#ac94e300de96c5e84492712644437fcb9
It is a tensor.

**Reply by Joseph-0123** [reply]
> See the doxygen for that adCoupledVectorGradient
> 
> ```
> const ADVectorVariableGradient & | adCoupledVectorGradient (const std::string &var_name, unsigned int comp=0) const
> ```
> 
> Then clicking on the return type: https://mooseframework.inl.gov/docs/doxygen/moose/MooseTypes_8h.html#ac94e300de96c5e84492712644437fcb9 It is a tensor.


Good, Guillaume. Thank you.  

Today I learn the [navier_stokes](https://github.com/idaholab/moose/tree/next/modules/navier_stokes). I find that there are two methods used to write the kernels for incompressible and compressible fluids in the navier_stokes module.

1) Using ADKernels. In the related ADkernels, the velocity is a vector and is not defined as 3 components.  The strong form residuals are defined as properties in the Materials and are directly used in the related ADKernels.

2) Using Kernels. In the related kernels, the velocity is a vector and defined as 3 components.  Each component is a scalar. The weak form residuals are written in the related Kernels.


I find that the incompressible fluid has the above two methods, but the compressible fluid only has the Kernel method. So I have questions about them.

1) InCompressible fluid

(1) Is there some difference between the two methods used for incompressible fluid besides their intrinsic differences (e.g., the ADkernel doesn't need the hand-coded Jacobian, etc)?  

(2) How could I choose them? Is the ADkernels with strong form residuals better than the Kernels with weak form residuals? 


2) Compressible fluid 

Can the ADKernels method with strong form residual be used for compressible fluid? How could I choose between the strong form residual and weak form residual if I use ADkernel method?

Thanks for your help.

J

**Reply by GiudGiud** [reply]
So whether a vector or 3 components are used is completely an implementation choice. We generally prefer using vectors for velocity, but it required more work.

Anything AD does not need to use a hand coded jacobian. That is the point of automatic differentiation, to be able to compute the derivatives of the residual to form the Jacobian. 

ADKernels still use the weak form. What you are seeing creating the strong form residual is for the purpose of SUPG stabilization. The weak form is still used for most of the solve

I recommend using ADKernels and implenenting the weak form

**Reply by Joseph-0123** [reply]
> We generally prefer using vectors for velocity, but it required more work.

Thanks, Guillaume.  I am using the ADKernels and implementing the weak form for my problem.

'We generally prefer using vectors for velocity, but it required more work.'  Sorry, I don't understand what 'more work' means here.

J

**Reply by GiudGiud** [reply]
In MOOSE we started by developing regular field variables. Vector variables came later and needed more development to work.

**Reply by Joseph-0123** [reply]
> In MOOSE we started by developing regular field variables. Vector variables came later and needed more development to work.

Hello, Guillaume. I am learning the [NSMomentumInviscidFlux.C
](https://github.com/idaholab/moose/blob/next/modules/navier_stokes/src/kernels/NSMomentumInviscidFlux.C). I find that the following content. 

Real
NSMomentumInviscidFlux::computeQpResidual()
{
  // For _component = k,

  // (rho*u) * u_k = (rho*u_k) * u <- we write it this way
  RealVectorValue vec(_u[_qp] * _u_vel[_qp],  // (U_k) * u_1
                      _u[_qp] * _v_vel[_qp],  // (U_k) * u_2
                      _u[_qp] * _w_vel[_qp]); // (U_k) * u_3

  // (rho*u_k) * u + e_k * P [ e_k = unit vector in k-direction ]
  vec(_component) += _pressure[_qp];

  // -((rho*u_k) * u + e_k * P) * grad(phi)
  return -(vec * _grad_test[_i][_qp]);
}


For example, if the component is 0, it means that _u represents velocity in the x direction. In this case, we will have _u*u_vel.  In my understanding, _u and u_vel are the same but here moose will only detect _u as a variable, right? This is strange to me since u_vel is also a variable that we need to solve. how to understand this? Thank you.

J

**Reply by GiudGiud** [reply]
Hello

I think _u is rho * v_x here.
This would be because the conserved variables (density, momentum, energy) are used for this set of kernels.

Guillaume

**Reply by Joseph-0123** [reply]
> Hello
> 
> I think _u is rho * v_x here. This would be because the conserved variables (density, momentum, energy) are used for this set of kernels.
> 
> Guillaume

yes, I know this. but u_vel is also _u, why do they have different expression and how moose can recognise that both are the same variable?

**Reply by GiudGiud** [reply]
u_vel is just the velocity not the momentum

**Reply by Joseph-0123** [reply]
> u_vel is just the velocity not the momentum

So u * v_vel is  rho * v_vel * v_vel.  

1. Can I use v_vel as the objective variable and rho * (u)^2 as the residual in my kernel?

2. What is the advantage of using  (rho * u_vel) as  objective variable (u) than v_vel as  objective variable ?

**Reply by GiudGiud** [reply]
1) Yes you can do that.
2) It depends on the problem at hand but some variable sets will lead to better conditioned problems. This is a little outside my expertise, I'd recommend you examine the literature.
